Mathlib/CategoryTheory/Groupoid/FreeGroupoidOfCategory.lean

```diff
@@ -0,0 +1,222 @@
+`C ⥤ Quotient (FreeGroupoid.homRel C)`. -/
+protected def FreeGroupoid := Quotient (FreeGroupoid.homRel C)
+
+instance [Nonempty C] : Nonempty (Category.FreeGroupoid C) :=
+  ⟨Quotient.mk (Quotient.mk ((Paths.of _).obj (Classical.arbitrary C)))⟩
+
+instance : Groupoid (Category.FreeGroupoid C) :=
+  Quotient.groupoid (Category.FreeGroupoid.homRel C)
+
+namespace FreeGroupoid
+
+/-- The localization map from the category `C` to the groupoid `Category.FreeGroupoid C` -/
+def of : C ⥤ Category.FreeGroupoid C where
+  __ := Quiver.FreeGroupoid.of C ⋙q (Quotient.functor (FreeGroupoid.homRel C)).toPrefunctor
+  map_id X := Quotient.sound _ (Category.FreeGroupoid.homRel.map_id X)
+  map_comp f g := Quotient.sound _ (Category.FreeGroupoid.homRel.map_comp f g)
+
+variable {C}
+
+lemma of_obj_bijective : Function.Bijective (of C).obj where
+  left _ _ h := by cases h; rfl
+  right X := ⟨X.as.as, rfl⟩
+
+section UniversalProperty
+
+variable {G : Type u₁} [Groupoid.{v₁} G]
+
+/-- The lift of a functor from `C` to a groupoid to a functor from
+`FreeGroupoid C` to the groupoid -/
+def lift (φ : C ⥤ G) : Category.FreeGroupoid C ⥤ G :=
+  Quotient.lift (FreeGroupoid.homRel C) (Quiver.FreeGroupoid.lift φ.toPrefunctor)
+    (fun _ _ f g r ↦ by
+      have {X Y : C} (f : X ⟶ Y) :=
+        Prefunctor.congr_hom (Quiver.FreeGroupoid.lift_spec φ.toPrefunctor) f
+      induction r <;> cat_disch)
+
+theorem lift_spec (φ : C ⥤ G) : of C ⋙ lift φ = φ :=
+  Functor.toPrefunctor_injective (by
+    change Quiver.FreeGroupoid.of C ⋙q
+      (Quotient.functor (FreeGroupoid.homRel C)).toPrefunctor ⋙q
+        (lift φ).toPrefunctor = φ.toPrefunctor
+    simp [lift, Quotient.lift_spec, Quiver.FreeGroupoid.lift_spec])
+
+theorem lift_unique (φ : C ⥤ G) (Φ : Category.FreeGroupoid C ⥤ G) (hΦ : of C ⋙ Φ = φ) :
+    Φ = lift φ := by
+  apply Quotient.lift_unique
+  apply Quiver.FreeGroupoid.lift_unique
+  exact congr_arg Functor.toPrefunctor hΦ
+
+theorem lift_comp {H : Type u₂} [Groupoid.{v₂} H] (φ : C ⥤ G) (ψ : G ⥤ H) :
+    lift (φ ⋙ ψ) = lift φ ⋙ ψ := by
+  symm
+  apply lift_unique
+  rw [← Functor.assoc, lift_spec]
+
+/-- The universal property of the free groupoid. -/
+def strictUniversalPropertyFixedTarget :
+    Localization.StrictUniversalPropertyFixedTarget (of C) ⊤ G where
+  inverts _ := inferInstance
+  lift F _ := lift F
+  fac _ _ := lift_spec ..
+  uniq F G h := by rw [lift_unique (of C ⋙ G) F h, ← lift_unique (of C ⋙ G) G rfl]
+
+attribute [local instance] Localization.groupoid
+
+instance : (of C).IsLocalization ⊤ :=
+  .mk' _ _ strictUniversalPropertyFixedTarget strictUniversalPropertyFixedTarget
+
+end UniversalProperty
+
+section Functoriality
+
+variable {D : Type u₁} [Category.{v₁} D] {E : Type u₂} [Category.{v₂} E]
+
+/-- The functor of free groupoid induced by a functor between the original categories. -/
+def map (φ : C ⥤ D) : Category.FreeGroupoid C ⥤ Category.FreeGroupoid D :=
```

Comment:
It would be nice to introduce constructors for objects and morphisms in the free groupoid:
```lean
abbrev mk (X : C) : Category.FreeGroupoid C := (of C).obj X
abbrev homMk {X Y : C} (f : X ⟶ Y) : mk X ⟶ mk Y := (of C).map f
```
so that you can have these simp lemmas:
```lean
@[simp]
lemma map_obj_mk (φ : C ⥤ D) (X : C) : (map φ).obj (mk X) = mk (φ.obj X) := rfl

@[simp]
lemma map_map_homMk (φ : C ⥤ D) {X Y : C} (f : X ⟶ Y) :
    (map φ).map (homMk f) = homMk (φ.map f) := rfl
```
(Similar lemmas should also be introduced for `lift`.)